In [ ]:
import pandas as pd
import matplotlib.pyplot as plt

## Предобработка

Поля датасета: 

* is_pinned - Закреплен ли пост - (0, 1)
* comments - Словаь информации о комменатриях, содержащий информацию о возможности комментировать, смотреть и количество комментариев
* marked_as_ads - Отмечен ли пост как рекламный - (0, 1)
* attachments - Вложения в посте (фото, видео, ссылки) - список
* date - Дата публикации поста - таймстемп
* from_id - ID автора поста - int
* id_x - Уникальный идентификатор поста - int
* likes - Информация по лайкам поста (аналогично комментарием)
* reactions - Информация о всех реакциях поста (словарь или список)
* owner_id - ID владельца поста - дублирует from_id
* post_source - Источник поста (например, API, web, mobile) - словарь с полем type
* reposts - Количество репостов поста - словарь, где количество лежит в поле count
* text - Текст поста - str
* views - Количество просмотров поста - словарь, где количество лежит в поле count
* donut_miniapp_url - Ссылка на мини-приложение, если пост связан с VK Donut - str
* check_sign - наличие подписи - bool
* id_y - Дублирующий идентификатор, связанный с группой - int
* description - Описание сообщества, в котором опубликован пост - str
* members_count - Количество подписчиков в сообществе - int
* activity - Рубрика сообщества - str
* status - Статус сообщества - str
* site - Сайт сообщества - str
* name - Название сообщества - str
* screen_name - Короткое название сообщества (URL-адрес) - str
* verified - Проверено ли сообщество (есть ли галочка) - (0, 1)



In [119]:
df = pd.read_csv('csv_files/df_wall.csv')
display(df.head())

,is_pinned,comments,marked_as_ads,attachments,date,from_id,id_x,likes,reactions,owner_id,...,check_sign,id_y,description,members_count,activity,status,site,name,screen_name,verified
0,1.0,"{'can_post': 1, 'can_view': 1, 'count': 74}",0,"[{'type': 'video', 'video': {'response_type': ...",1740042002,-211229778,49105,"{'can_like': 1, 'count': 3895, 'user_likes': 0...","[{'id': 0, 'title': 'Лайк', 'asset': {'animati...",211229778,...,NaN,211229778,"Шоу, в котором четыре девушки дают советы по л...",706762,"Шоу, передача",NaN,https://www.youtube.com/channel/UCXDerjAwOSH30...,Женский Форум,zhenforum,1
1,NaN,"{'can_post': 1, 'can_view': 1, 'count': 2}",0,"[{'type': 'video', 'video': {'response_type': ...",1741093583,-211229778,49839,"{'can_like': 1, 'count': 1056, 'user_likes': 0...","{'count': 1056, 'items': [{'id': 0, 'count': 1...",211229778,...,NaN,211229778,"Шоу, в котором четыре девушки дают советы по л...",706762,"Шоу, передача",NaN,https://www.youtube.com/channel/UCXDerjAwOSH30...,Женский Форум,zhenforum,1
2,NaN,"{'can_post': 1, 'can_view': 1, 'count': 6}",0,"[{'type': 'video', 'video': {'response_type': ...",1741007166,-211229778,49794,"{'can_like': 1, 'count': 1022, 'user_likes': 0...","{'count': 1022, 'items': [{'id': 0, 'count': 1...",211229778,...,NaN,211229778,"Шоу, в котором четыре девушки дают советы по л...",706762,"Шоу, передача",NaN,https://www.youtube.com/channel/UCXDerjAwOSH30...,Женский Форум,zhenforum,1
3,NaN,"{'can_post': 1, 'can_view': 1, 'count': 7}",0,"[{'type': 'photo', 'photo': {'album_id': -7, '...",1740994531,-211229778,49786,"{'can_like': 1, 'count': 275, 'user_likes': 0,...","{'count': 275, 'items': [{'id': 0, 'count': 27...",211229778,...,NaN,211229778,"Шоу, в котором четыре девушки дают советы по л...",706762,"Шоу, передача",NaN,https://www.youtube.com/channel/UCXDerjAwOSH30...,Женский Форум,zhenforum,1
4,NaN,"{'can_post': 1, 'can_view': 1, 'count': 39}",0,"[{'type': 'photo', 'photo': {'album_id': -7, '...",1740665284,-211229778,49617,"{'can_like': 1, 'count': 4967, 'user_likes': 0...","{'count': 4967, 'items': [{'id': 0, 'count': 4...",211229778,...,NaN,211229778,"Шоу, в котором четыре девушки дают советы по л...",706762,"Шоу, передача",NaN,https://www.youtube.com/channel/UCXDerjAwOSH30...,Женский Форум,zhenforum,1


Размер полученного датасета

In [120]:
df.shape

(2626, 25)

Собрано 2626 постов, сейчас есть 25 полей, в которых содержится информация как о самом посте, так и сообществе, который опубликовал этот пост

По описанию полей видно, что много полей хранятся в неудобном виде. Приведем в порядок

None  в is_pinned по специфике сборки через АПИ означает, что пост не закреплен, поэтому заменим значения на 0

In [121]:
df['is_pinned'].fillna(0, inplace=True)
df['is_pinned'] = df['is_pinned'].astype(int)

/var/folders/wj/jzwd6lfn7xx02ns8v_wr29bm0000gn/T/ipykernel_91931/511727329.py:1: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df['is_pinned'].fillna(0, inplace=True)


В поле о комментах записана инфа о возможности комментирования, просмотра поста и количества комментов. Выделим эту инфу в разные поля

In [122]:
df['can_comment'] = df['comments'].apply(lambda x: eval(x).get('can_post')) 
df['can_view'] = df['comments'].apply(lambda x: eval(x).get('can_view'))
df['comments_cnt'] = df['comments'].apply(lambda x: eval(x).get('count'))

В поле вложений лежит список всех вложений с кучей нформацией лишней, достанем только уникальные типы вложений поста

In [123]:
df['attachments_types'] = df['attachments'].apply(lambda x: set({item['type'] for item in eval(x)}))

Перевед дату в тип даты

In [124]:
df['date'] = pd.to_datetime(df['date'], unit='s')

По аналогии с комментариями раскрываем лайки

In [126]:
df['can_like'] = df['likes'].apply(lambda x: eval(x).get('can_like')) 
df['can_repost'] = df['likes'].apply(lambda x: int(not eval(x).get('repost_disabled')))
df['likes_cnt'] = df['likes'].apply(lambda x: eval(x).get('count'))

Можно заметить, что поле с реакциями имеет разный формат, тут придется помучиться, чтобы хорошо преобразовать

In [127]:
reaction_titles = {
    0: 'Лайк',
    1: 'Смешно',
    2: 'Ого!',
    3: 'Восторг',
    4: 'Печаль',
    5: '&$#%!'
}

def parse_reactions(row):
    reactions = eval(row['reactions'])
    result = {}

    if isinstance(reactions, list):
        result['Лайк'] = row['likes_cnt']

    elif isinstance(reactions, dict):
        items = reactions.get('items', [])
        for item in items:
            reaction_id = item.get('id')
            title = reaction_titles.get(reaction_id, 'Неизвестная реакция')
            count = item.get('count', 0)
            result[title] = count

    return result

df['reactions_types'] = df.apply(parse_reactions, axis=1)


Теперь сделаем имитацию OHE

In [128]:
for _, reaction  in reaction_titles.items():
    df[reaction] = df['reactions_types'].apply(lambda x: x.get(reaction, 0))

Далее приводятся в нормальный вид похожие по структуре поля (вытаскиваем колчиество или название)

In [134]:
df['source'] = df['post_source'].apply(lambda x: eval(x).get('type'))
df['reposts_cnt'] = df['reposts'].apply(lambda x: eval(x).get('count'))
df['views_cnt'] = df['views'].apply(lambda x: eval(x).get('count'))

donut_miniapp_url - сама ссылка нам не нужна, но будем хранить информацию о том, пользуется ли сообщество такой штукой

In [156]:
df['is_donut'] = df['donut_miniapp_url'].apply(lambda x: 0 if pd.isna(x) else 1)

Что-то похожее с check_sign

In [ ]:
df['is_sign'] = df['check_sign'].apply(lambda x: 0 if pd.isna(x) else 1)

Статус сообщества, Сайт сообщества сами не нужны, а вот с информацией есть ли они было бы интересно ознакомиться

In [ ]:
df['is_status'] = df['status'].apply(lambda x: 0 if pd.isna(x) else 1)
df['is_site'] = df['site'].apply(lambda x: 0 if pd.isna(x) else 1)

Удаляем айдишники, так как для анализа они не нужны, а имена сообществ у нас уникальные сами по себе, также удалим короткое название сообщества и поля, которые мы преобразовали

In [164]:
df.drop(columns=['comments' , 'attachments'      , 'from_id', 'id_x'     , 'owner_id'   , 
                 'id_y'     , 'screen_name'      , 'likes'  , 'likes_cnt', 'post_source',
                 'reactions', 'reactions_types'  , 'reposts', 'views'    , 'check_sign' ,
                 'status'   , 'donut_miniapp_url', 'site'   ,], inplace=True)

In [166]:
df.head()

,is_pinned,marked_as_ads,date,text,description,members_count,activity,name,verified,can_comment,...,Лайк,Смешно,Ого!,Восторг,Печаль,&$#%!,reposts_cnt,source,views_cnt,is_donut
0,1,0,2025-02-20 09:00:02,Можно бесконечно смотреть на три явления: как ...,"Шоу, в котором четыре девушки дают советы по л...",706762,"Шоу, передача",Женский Форум,1,1,...,3895,0,0,0,0,0,2708,vk,727617,0
1,0,0,2025-03-04 13:06:23,NaN,"Шоу, в котором четыре девушки дают советы по л...",706762,"Шоу, передача",Женский Форум,1,1,...,1052,4,0,0,0,0,87,vk,99412,0
2,0,0,2025-03-03 13:06:06,NaN,"Шоу, в котором четыре девушки дают советы по л...",706762,"Шоу, передача",Женский Форум,1,1,...,1017,5,0,0,0,0,40,vk,95628,0
3,0,0,2025-03-03 09:35:31,Сегодня в 15:00 (мск) откроем продажу билетов ...,"Шоу, в котором четыре девушки дают советы по л...",706762,"Шоу, передача",Женский Форум,1,1,...,274,0,0,1,0,0,13,api,105038,0
4,0,0,2025-02-27 14:08:04,✍ Ты медленно пишешь заявление на увольнение...,"Шоу, в котором четыре девушки дают советы по л...",706762,"Шоу, передача",Женский Форум,1,1,...,4956,10,0,0,0,1,1083,api,560361,0


Итого с пустыми значениями остались описание сообщества и текст, так как это строки, то сюда можно подставить пустую строку - то есть в описании или посте пусто, нет нарушений

In [169]:
df.isnull().sum()


is_pinned              0
marked_as_ads          0
date                   0
text                 790
description          101
members_count          0
activity               0
name                   0
verified               0
can_comment            0
can_view               0
comments_cnt           0
attachments_types      0
can_like               0
can_repost             0
Лайк                   0
Смешно                 0
Ого!                   0
Восторг                0
Печаль                 0
&$#%!                  0
reposts_cnt            0
source                 0
views_cnt              0
is_donut               0
dtype: int64

In [171]:
df['text'].fillna("", inplace=True)
df['description'].fillna("", inplace=True)

/var/folders/wj/jzwd6lfn7xx02ns8v_wr29bm0000gn/T/ipykernel_91931/2947034952.py:2: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df['description'].fillna("", inplace=True)


In [172]:
df.isnull().sum()

is_pinned            0
marked_as_ads        0
date                 0
text                 0
description          0
members_count        0
activity             0
name                 0
verified             0
can_comment          0
can_view             0
comments_cnt         0
attachments_types    0
can_like             0
can_repost           0
Лайк                 0
Смешно               0
Ого!                 0
Восторг              0
Печаль               0
&$#%!                0
reposts_cnt          0
source               0
views_cnt            0
is_donut             0
dtype: int64

Ура!)

## Анализ